### PR TITLE
BEAM-13189 Python TextIO: add escapechar feature.

### PR DESCRIPTION
**Pull request by nikie**
Add `escapechar` parameter to Python TextIO transforms: `ReadFromText`,  `ReadAllFromText`:
* escaped line delimiters are skipped when splitting, for example: `'aa\\\naa\n'` would result in 1 line `'aa\\\naa'`
* this is for line splitting only, so escapechars are not removed from the lines, they are supposed to be parsed separately, for example with `csv` library
* escaped escape chars are also handled, so that if there are odd number of escape chars before line delimiter - it will be escaped, but if there are even number of escape chars - all are treated as escaped themselves and the line delimiter will be used for splitting
* using `bytes` type for the `escapechar` in line with the custom delimiter feature, but keeping name `escapechar` as in the `csv` library (arguable)
* only 1 byte `escapechar` length is supported (in line with `csv`)

Closes: BEAM-13189

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by nikie** [comment]
R: @dmitriikuzinepam 
R: @tvalentyn 
Unfortunately this adds even more `if/else` blocks for each default case line splitting which may affect performance.

In this regard, I can propose to move custom delimiter and escapechar features handling into a separate `_find_separator_bounds` method, return the old one (before these features) and choose between them in the init based on the source params. This would result in code duplication, but the old method was quite short and as long as all cases are covered with unit tests, this could be considered...

**Comment by tvalentyn** [comment]
Run Python 3.8 PostCommit

**Comment by tvalentyn** [comment]
(I was testing jenkins trigger commands and used this PR as an example. Looks like the trigger commands are  not working right now.)

**Comment by nikie** [comment]
> (I was testing jenkins trigger commands and used this PR as an example. Looks like the trigger commands are not working right now.)

@tvalentyn 
The hook started at 17:29:34 UTC, 11 minutes after your commit
https://ci-beam.apache.org/job/beam_PostCommit_Python38_PR/350/console

File: sdks/python/apache_beam/io/textio.py [hunk 1]
```diff
@@ -362,6 +391,15 @@ def _is_self_overlapping(delimiter):
         return True
     return False
 
+  def _is_escaped(self, read_buffer, position):
+    # Returns True if byte at position is preceded with an odd number
+    # of escapechar bytes or False if preceded by 0 or even escapes
+    # (the even number means that all the escapes are escaped themselves).
+    for current_pos in reversed(range(-1, position)):
+      if read_buffer.data[current_pos:current_pos + 1] != self._escapechar:
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/textio.py:399]: This logic is a little tricky to follow:
1. why do we go to -1 ?
2. can we compare a single byte at a position instead of a 1-character range?

why not to count `preceding_escape_chars` in a loop and then  `return preceding_escape_chars % 2 == 1` ?

Comment (nikie) [reply sdks/python/apache_beam/io/textio.py:399]: Totally agree about the trickiness. But it is beautiful and works :) 
I will look for more explicit solution if you insist, though.

1. -1 is needed to check if the 1st character is escaped without additional check if it is the first or not and be able to return from the same if block if we reached the start of buffer, -1:0 range returns empty bytes which cannot be equal to escapechar as we do not allow empty one:
```
>>> b'\\'[-1:0]
b''
>>> b'\\'[-1:0] == b'\\'
False
```
2. No, we cannot compare `byte` to `bytes` (I was a bit surprised as well, but we are doing similar comparisons in `_find_separator_bounds`):
```
>>> b'a'[0] == b'a'
False
>>> b'a'[0:1] == b'a'
True
```

Comment (nikie) [reply sdks/python/apache_beam/io/textio.py:399]: It looks like the last line `return False` cannot be reached, because we loop until first non-escape or the start of buffer and always return. That's the problem...
It only passes Pylint validation because `inconsistent-return-statements` is disabled, but can confuse readers.

Comment (tvalentyn) [reply sdks/python/apache_beam/io/textio.py:399]: Thanks. 
I personally like `explicit is better than implicit` from https://www.python.org/dev/peps/pep-0020/ . Someone reading this code might wonder whether [-1:0] range was an intentional choice to test implicitly test edge case, or it's an off-by-one error.

Comment (nikie) [reply sdks/python/apache_beam/io/textio.py:400]: I have updated the code to use an explicit counter. Now it is easier to understand.

Should I replace bytes slicing comparison with `if read_buffer.data[current_pos] != self._escapechar[0]`, or let's keep it consistent with `_find_separator_bounds`?

What about my suggestion in the first comment to have 2 `_find_separator_bounds` - one for the default case, another for custom delimiter and/or escapechar to avoid extra ifs in the default use case?

I can prepare such version to see how it would look, if you are ready to consider such code duplication for performance reasons.

Comment (tvalentyn) [reply sdks/python/apache_beam/io/textio.py:400]: I doubt that there is a significant performance impact, but you could try to run https://github.com/apache/beam/blob/9a44c9e6e1d944cc022d3e3e11b7ec08f7a3a790/.test-infra/jenkins/job_PerformanceTests_Python.groovy#L39

**Review by tvalentyn — APPROVED** [review]

**Review by dmitriikuzinepam — APPROVED** [review]

**Comment by nikie** [comment]
Run Python37 WordCountIT Performance Test

**Comment by tvalentyn** [comment]
Looks like a trigger command only works while a PR is open. But we run these tests continuously, see: http://metrics.beam.apache.org. Specifically
http://metrics.beam.apache.org/d/1cnwVDkGk/python-wordcount-it-benchmarks?orgId=1

**Comment by nikie** [comment]
> Looks like a trigger command only works while a PR is open. But we run these tests continuously, see: 
Also available from Jenkins:
https://ci-beam.apache.org/view/PerformanceTests/job/beam_PerformanceTests_WordCountIT_Py37/
The latest run today after the merge took 14 min: https://ci-beam.apache.org/view/PerformanceTests/job/beam_PerformanceTests_WordCountIT_Py37/3590/testReport/
which is similar to previous runs, including the ones before the custom delimiter feature.
